### PR TITLE
[client/network] ban peer for 10s after disconnecting

### PR DESCRIPTION
**Pull request by melekes**
plus do not accept incoming connections if peer is still banned.

Refs https://github.com/paritytech/infrastructure-alerts/issues/26
Refs #13778

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2609652

**Comment by melekes** [comment]
@altonen @dmitry-markin this PR solves the problem from "our" side by ignoring incoming connections which arrive too soon (peers are banned for 10s).

**Comment by bkchr** [comment]
> plus do not accept incoming connections if peer is still banned.

Is this actually on a connection level? Isn't this just on the "substream" level? 

Any possibility to add some simple test? 

When do we ban a peer? Only when they lost all their reputation? Or would we also ban them in other circumstances?

**Comment by bkchr** [comment]
And the peerset already has some kind of banning or? When a node "runs out of reputation" we already had some banning behavior, but 30 seconds AFAIR or?

**Comment by bkchr** [comment]
> And the peerset already has some kind of banning or? When a node "runs out of reputation" we already had some banning behavior, but 30 seconds AFAIR or?

https://github.com/paritytech/substrate/blob/ee38a52d9ea643e4eec1e954b0b603aea4461416/client/peerset/src/lib.rs#L492-L493 ahh that was the number I was speaking about, but this is only the time it takes to change the reputation by half.

**Comment by dmitry-markin** [comment]
What was the reason why this wasn't merged last time?

**Comment by dmitry-markin** [comment]
> Fixes https://github.com/paritytech/substrate/issues/13778

I wouldn't tell that this fixes the referenced issue, it only puts misbehaving peers "on hold" locally.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2639704

**Comment by melekes** [comment]
> the underlying issue for this behavior has to be found out first, otherwise we're just masking it with this change.

I agree ☝️ 

> Is this actually on a connection level? Isn't this just on the "substream" level?

All the streams within the notifications protocol.

> When do we ban a peer?

Any time we choose to disconnect from a peer (bad handshake, max peers reached).

**Comment by bkchr** [comment]
> Any time we choose to disconnect from a peer (bad handshake, max peers reached).

Do we also ban when the peer has less than the minimum reputation in any peerset?

**Comment by melekes** [comment]
> Do we also ban when the peer has less than the minimum reputation in any peerset?

Yes, but that happens independently in the peerset afaik.

File: client/network/src/protocol/notifications/behaviour.rs [hunk 1]
```diff
@@ -1533,6 +1552,12 @@ impl NetworkBehaviour for Notifications {
 
 					// Disabled => Disabled | Incoming
 					PeerState::Disabled { mut connections, backoff_until } => {
+						// Do not accept incoming connections if the peer is still banned.
+						if backoff_until.map_or(false, |t| t > Instant::now()) {
+							*entry.into_mut() = PeerState::Disabled { connections, backoff_until };
```
[the remark this answers is not in the document]
Comment (melekes) [reply client/network/src/protocol/notifications/behaviour.rs:1557]: 🤔 I was under the impression that the peer will try again at some point in the future (i.e. the same block will be executed again => progress is made).

**Comment by melekes** [comment]
> but the underlying issue for this behavior has to be found out first

Closing this until we find out the reason for peers constantly reconnecting to us (see https://github.com/paritytech/substrate/issues/13778#issuecomment-1499140724 for @altonen's thoughts).

**Pull request closed without merging**